### DISCUSSION TITLE
Plot field variable on IPF contour

### DISCUSSION THREAD

**Discussion by arovinelli** [pyxem/orix#440] [Q&A]
hi there, very nice tool.
I have a quick question: I would like to plot a customs scalar variable (say porosity)  over an IPF, starting from discrete orientation data.  The plot should be filled (same as [here](https://orix.readthedocs.io/en/stable/examples/inverse_pole_figures/inverse_pole_density_function.html#sphx-glr-examples-inverse-pole-figures-inverse-pole-density-function-py) ), and the color should be associated to the custom variable value. Is this currently possible ?

Thanks in advance

**Comment by hakonanes** [comment]
Hi @arovinelli,

thanks! Assuming you have an auxiliary value (porosity or similar) for each data point (orientation), you can pass these to the `weights` parameter, as described in the documentation of the [pole_density_function()](https://orix.readthedocs.io/en/stable/reference/generated/orix.plot.InversePoleFigurePlot.pole_density_function.html). Have you tried this?

**Reply by arovinelli** [reply]
Hi @hakonanes , thanks for your reply.
I tried what you suggested it but it does not do what I need.
Using `weights`  scale the pole density function field based on weight. What I want is the plot of auxiliary variable itself,  similar to what you would do when plotting the Schmid factor on the IPF (see [here](https://groups.google.com/g/mtexmail/c/OXG-sjFRyRU) for an example).

**Reply by hakonanes** [reply]
I see, thanks for clarifying, and sorry for my misunderstanding.

What you're asking for is not implemented, unfortunately. @harripj wrote the pole density function functionality, he might have some idea for where to start to add this feature.

**Comment by harripj** [comment]
Hi @arovinelli,

this isn't easily doable from the public facing API but can be done by following the logic in [orix.measure.pole_density_function](https://github.com/pyxem/orix/blob/87fb7cb0c9b75129570badf6625db9fc97a14477/orix/measure/pole_density_function.py#L29) and adapting for your use case.

Here's a small example which may help:
```python
from orix.sampling.S2_sampling import _sample_S2_equal_area_coordinates
from orix.plot import StereographicPlot
from orix.projections import StereographicProjection
from orix.vector import Vector3d
import numpy as np
from matplotlib import pyplot as plt

# setup projection
hemisphere = "upper"

poles = {"upper": -1, "lower": 1}
sp = StereographicProjection(poles[hemisphere])

# setup polar grid
resolution = 1.0  # degrees
azimuth_coords, polar_coords = _sample_S2_equal_area_coordinates(
    resolution,
    hemisphere=hemisphere,
    azimuth_endpoint=True,
)

azimuth_grid, polar_grid = np.meshgrid(azimuth_coords, polar_coords, indexing="ij")

# simulate auxiliary variable
azimuth_center_grid, polar_center_grid = np.meshgrid(
    azimuth_coords[:-1] + np.diff(azimuth_coords) / 2,
    polar_coords[:-1] + np.diff(polar_coords) / 2,
    indexing="ij",
)

variable = np.zeros_like(polar_center_grid)
mask = (azimuth_center_grid >= 1 * np.pi / 3) * (azimuth_center_grid < 2 * np.pi / 3)
variable[mask] = np.cos(polar_center_grid[mask])

mask = (azimuth_center_grid >= 4 * np.pi / 3) * (azimuth_center_grid < 5 * np.pi / 3) 
variable[mask] = np.cos(polar_center_grid[mask])

# plot
fig, ax = plt.subplots(subplot_kw=dict(projection="stereographic"))

v_grid = Vector3d.from_polar(azimuth=azimuth_grid, polar=polar_grid).unit
x, y = sp.vector2xy(v_grid)

ax.pcolormesh(x.reshape(azimuth_grid.shape), y.reshape(azimuth_grid.shape), variable)
```
<img width="232" alt="0a1a417d-b4d8-4467-8105-8660a0a29bf9" src="https://user-images.githubusercontent.com/16853829/228672079-04c1ef83-f7e6-4dff-957c-8d117aa926f9.png">


```python
from orix.quaternion.symmetry import D3 as symmetry

# in crystal fundamental sector
v_center_grid = Vector3d.from_polar(
    azimuth=azimuth_center_grid, polar=polar_center_grid
).unit
# fold back in into fundamental sector
v_center_grid_fs = v_center_grid.in_fundamental_sector(symmetry)
azimuth_center_fs, polar_center_fs, _ = v_center_grid_fs.to_polar()

# calculate histogram values for vectors folded back into
# fundamental sector
i = np.digitize(azimuth_center_fs, azimuth_coords[1:-1])
j = np.digitize(polar_center_fs, polar_coords[1:-1])
# recompute histogram
temp = np.zeros((azimuth_coords.size - 1, polar_coords.size - 1))
# add hist data to new histogram without buffering
np.add.at(temp, (i.ravel(), j.ravel()), variable.ravel())

temp = np.ma.array(temp, mask=~(v_center_grid <= symmetry.fundamental_sector))

# plot
fig, ax = plt.subplots(subplot_kw=dict(projection="stereographic"))

v_grid = Vector3d.from_polar(azimuth=azimuth_grid, polar=polar_grid).unit
x, y = sp.vector2xy(v_grid)

ax.pcolormesh(x.reshape(azimuth_grid.shape), y.reshape(azimuth_grid.shape), temp)
ax.restrict_to_sector(symmetry.fundamental_sector)
```
<img width="338" alt="ce470f55-2ef5-4c98-9cfd-8a521ca507f8" src="https://user-images.githubusercontent.com/16853829/228672329-c741dcb8-48b4-4fe6-8e8e-0f21bf93b56f.png">


If you are looking for a contour plot like in the reference you cited then this is possible using `plt.contour` or `pet.contourf`, for example:
```python
v_grid = Vector3d.from_polar(azimuth=azimuth_center_grid, polar=polar_center_grid).unit
x, y = sp.vector2xy(v_grid)

ax.contourf(x.reshape(v_grid.shape), y.reshape(v_grid.shape), temp, levels=10)
ax.restrict_to_sector(symmetry.fundamental_sector)
```
<img width="338" alt="17b441fe-9945-4d80-86ed-2fe53e5accd4" src="https://user-images.githubusercontent.com/16853829/228672375-73ba834a-623e-4afc-8c2e-2473c5e9c88b.png">

I hope this is of use and points you in the right direction! If you think this feature is of general use then we could consider adding a simpler public facing API in the future.

**Reply by arovinelli** [reply]
hi @harripj thanks a lot for the examples this is exactly what I need, however there is a catch.
I have access to a limited dataset (~50 orientations) which does not necessarily covers all the fundamental space.
my dataset looks like this:
euler1, euler2, euler3, var1, var2, ...
...
...
...

I would need to interpolate values on the fundamental region using azimutal coordinates but I'm unsure how to do this (maybe you know :) ).
i was able to run your example, but I was not able to adapt it to my case (i.e starting from few points)
In general I think that this capability would be helpful especially to correlate variables with orientation, however it should be wrapped around something that operates given a set of discrete points for example, plotFieldVariableIPF(euler, values, symmetry) and does all the stuff you did in the background.

**Reply by hakonanes** [reply]
@harripj, thank you for providing these examples!

We should definitely find a clean way to include them in the API. I suggest to
1. Add the first snippet as an example to the [stereographic projection examples](https://orix.readthedocs.io/en/stable/examples/index.html#stereographic-projection) and the second two to the [inverse pole figure examples](https://orix.readthedocs.io/en/stable/examples/index.html#inverse-pole-figures)
2. As methods to the classes [StereographicPlot](https://orix.readthedocs.io/en/stable/reference/generated/orix.plot.StereographicPlot.html) and [InversePoleFigure](https://orix.readthedocs.io/en/stable/reference/generated/orix.plot.InversePoleFigurePlot.html)
3. Finally, add convenience methods to the classes [Vector3d](https://orix.readthedocs.io/en/stable/reference/generated/orix.vector.Vector3d.html) and [Miller](https://orix.readthedocs.io/en/stable/reference/generated/orix.vector.Miller.html)

We should at least do 1. as soon as possible, and then 2. and 3. can wait till someone has time.

**Comment by harripj** [comment]
@arovinelli thanks for clarifying. If you wanted to plot some orientations in the Inverse Pole Figure, you could view for example the lab Z-direction in the crystal reference frame. I think the example below is perhaps more suited to your use case:

```python
from orix.measure import pole_density_function
from orix.vector import Vector3d
from orix.quaternion import Orientation
import numpy as np
from matplotlib import pyplot as plt
from orix.quaternion.symmetry import Th

symmetry = Th

# 10 random euler angles
euler = np.random.rand(10, 3) * 2 * np.pi
ori = Orientation.from_euler(euler)  # create orientations
v = ori * Vector3d.zvector()  # sample-Z in crystal reference frame

# values of variable for a given orientation
vals = np.arange(v.size) + 1  # 1, 2, 3 ...

hist, (x, y) = pole_density_function(
    v,
    resolution=0.5,
    sigma=0,
    log=False,
    hemisphere="upper",
    symmetry=symmetry,
    weights=vals,
    mrd=False,
)

# plot
fig, ax = plt.subplots(subplot_kw=dict(projection="ipf", symmetry=symmetry))

pc = ax.pcolormesh(x, y, hist)

plt.colorbar(pc, label="Variable", ax=ax)
```
<img width="306" alt="02c6f4df-b2ea-42ea-a414-3f9c8b5111fc" src="https://user-images.githubusercontent.com/16853829/229302613-26cfaa81-9b5d-4d73-9b42-2975e39cce5a.png">

You can see here that the values scale from 0-10 as expected from the setup (`vals`). Note that this is the case when `sigma=0, log=False, mrd=False`, otherwise the values will change due to blurring, logarithm, or scaled as Multiples of Random Density (MRD).

To demonstrate this you may wish to apply some blurring to improve the visualisation of the data or to simulate some uncertainty in the data. In the example below `sigma=2` degrees, but note that the scale changes.
<img width="316" alt="6596b868-daea-4924-8dfd-b9f5b656f157" src="https://user-images.githubusercontent.com/16853829/229303086-93c8b5dd-5f17-42a0-b5ef-ec22d7cbf3c9.png">

This scaling can be undone in this case by scaling as `hist = hist * vals.max() / hist.max()`
<img width="306" alt="8a01c3f4-a8b9-4d80-819e-a792aaf8b4a0" src="https://user-images.githubusercontent.com/16853829/229303194-5fc11bd0-32b6-4158-b196-0d2a3ea1f8f8.png">

**Comment by harripj** [comment]
@hakonanes yes I think there is some value in including them in the docs too. It would also be nice to add a option to plot the PDF as a contour plot (both filled and unfilled), as this seems to be a common representation in the field. I can look into this when I find some time 😄.
